README.md
# <a name="title"></a> Kitchen::Marathon

[![Gem Version](https://badge.fury.io/rb/kitchen-marathon.svg)](http://badge.fury.io/rb/kitchen-marathon)
[![Build Status](https://travis-ci.org/yieldbot/kitchen-marathon.svg?branch=master)](https://travis-ci.org/yieldbot/kitchen-marathon)

A Test Kitchen Driver for Mesos Marathon.

This driver uses the [Marathon REST API][marathon_api] to create and destroy Marathon applications which act as Kitchen suites, allowing you to leverage the resources of a Mesos cluster to help reduce testing time.

## <a name="requirements"></a> Requirements

### Mesos/Marathon

To use this driver you will need access to a Mesos Marathon cluster.  You will need to know the hostname and access credentials of the cluster for the plugin to access the REST API.

This cluster must be configured to run [Docker containers][marathon_docker].

### Marathon Job Configuration

This driver requires the following be true of the Marathon job configuration:

* You are running a Marathon container job of type "DOCKER".
* You can SSH into the container with a user specified username and private key.

This driver requires a slightly opinionated approach to Marathon Application configuration to assist in identifying what host port to use to acccess SSH running inside of a container.  

If there is a single port mapping defined, the plugin will assume that this is where SSH is listening, if there is more than one port mapping defined, the plugin looks for a mapping label with the key of "SERVICE" and a value of "SSH".

The following is a slice of json showing the label:

```json
{
  "container": {
    "docker": {
      "forcePullImage": true,
      "image": "someimagenamehere:latest",
      "network": "BRIDGE",
      "portMappings": [
        {
          "containerPort": 22,
          "hostPort": 0,
          "labels": {
            "SERVICE" : "ssh"
          },
          "protocol": "tcp"
        },
      ],
    },
    "type": "DOCKER",
  }
}

```

## Installation and Setup

Please read the Test Kitchen [docs][test_kitchen_docs] for more details.

Example `.kitchen.local.yml`:

```yaml
---
driver:
  name: marathon

# Username and key used to access the container
transport:
  ssh_key: '~/.ssh/vagrant'
  username: kitchen

driver_config:
  app_launch_timeout: 30
  app_template: '.kitchen-marathon.json'
  marathon_host: 'http://core-apps.mesos-marathon.service.consul:8080'

platforms:
- name: centos-7
  driver_config:
    app_config:
      container:
        docker:
          image: jdeathe/centos-ssh:centos-7
  run_list:
  - recipe[yum]
```

## <a name="config"></a> Configuration

### <a name="config-app-prefix"></a> app\_prefix

This is the prefix applied to Marathon App names.

The default value is `kitchen/`.

### <a name="config-app-template"></a> app\_template

This allows for a Marathon Job template that will be used as the base template when the job template is created.  If this value is not provided all configuration parameters need to be set via `app_config`.

The default value is `nil`.

### <a name="config-app-config"></a> app\_config

This allows for Marathon Job configuration to be configured at the platform/suite/etc level as necessary.

This configuration is merged on top of any configuration provided by `app_template`.

The default value is `{}`.

### <a name="config-app-launch-timeout"></a> app\_launch\_timeout

Determines the timeout, in seconds, that the driver will wait for a Marathon App to deploy.  If the timeout is reached, the driver will assume that the deployment will not succeed and will attempt to stop the deployment and delete the Marathon App.  (This cleanup is best effort.)

The default value is `30` seconds.

### <a name="config-marathon-host"></a> marathon\_host

The web address to the Marathon host.

The default value is `http://localhost:8080`.

### <a name="config-marathon-password"></a> marathon\_password

The password to use for an HTTP AUTH protected Marathon Host

The default value is `nil`.

### <a name="config-marathon-username"></a> marathon\_username

The username to use for an HTTP AUTH protected Marathon Host

The default value is `nil`.

### <a name="config-marathon-verify-ssl"></a> marathon\_verify\_ssl

Whether or not a certificate presented by the Marathon Host is verified.

The default value is `true`.

## Marathon Proxy Configuration

### <a name="config-marathon-proxy-address"></a> marathon\_proxy\_address

Provides the option to specify a proxy address if necessary when connecting to the Marathon host.

The default value is `nil`.

### <a name="config-marathon-proxy-password"></a> marathon\_proxy\_password

Provides the option to specify a proxy passowrd if necessary when connecting to the Marathon host.

The default value is `nil`.

### <a name="config-marathon-proxy-port"></a> marathon\_proxy\_port

Provides the option to specify a proxy port if necessary when connecting to the Marathon host.

The default value is `nil`.

### <a name="config-marathon-proxy-username"></a> marathon\_proxy\_username

Provides the option to specify a proxy username if necessary when connecting to the Marathon host.

The default value is `nil`.

## Development

* Source hosted at [GitHub][repo]
* Report issues/questions/feature requests on [GitHub Issues][issues]

Pull requests are very welcome! Make sure your patches are well tested.
Ideally create a topic branch for every separate change you make. For
example:

1. Fork the repo
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## Authors

Created and maintained by [Anthony Spring][author] (<aspring@yieldbot.com>)

## License

Apache 2.0 (see [LICENSE][license])

[author]:             https://github.com/yieldbot
[issues]:             https://github.com/yieldbot/kitchen-marathon/issues
[license]:            https://github.com/yieldbot/kitchen-marathon/blob/master/LICENSE
[marathon]:           https://mesosphere.github.io/marathon/
[marathon_api]:       https://github.com/otto-de/marathon-api
[marathon_docker]:    https://mesosphere.github.io/marathon/docs/native-docker.html
[repo]:               https://github.com/yieldbot/kitchen-marathon
[test_kitchen_docs]:  http://kitchen.ci/docs/getting-started/



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
- Nothing at this time.

## [0.1.0]
### Notes
- Initial public release

## [0.0.3]
### Changes
- Make the assumption that if a single port mapping is available that it is the SSH port.

## [0.0.2]
### Changes
- Remove unnecessary ssh configuration params -- this is handled by the transport.
- Renamed the proxy configuration pieces to be reflecting of being tied to marathon host.
- Fixed an issue with the merging of the 3 configuration pieces with regards to symbols and deep merging sanity.
- Changed the default prefix to include the forward slash -- making it optional.
- Fixed an issue with how app_id was managed in state that allows for apps to not be cleaned up properly.

# 0.0.1
### Added
- Initial internal release

[Unreleased]: https://github.com/yieldbot/kitchen-marathon/compare/kitchen-marathon-0.1.0...HEAD
[0.1.0]: https://github.com/yieldbot/kitchen-master/compare/kitchen-marathon-0.0.3...kitchen-marathon-0.1.0
[0.0.3]: https://github.com/yieldbot/kitchen-master/compare/kitchen-marathon-0.0.2...kitchen-marathon-0.0.3
[0.0.2]: https://github.com/yieldbot/kitchen-master/compare/kitchen-marathon-0.0.1...kitchen-marathon-0.0.2

lib/kitchen/driver/marathon_version.rb
# -*- encoding: utf-8 -*-
#
# Author:: Anthony Spring (<aspring@yieldbot.com>)
#
# Copyright (C) 2016, Anthony Spring
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'json'

module Kitchen
  module Driver
    # This defines the version of the gem
    module Version
      MAJOR = 0
      MINOR = 1
      PATCH = 0
      BUILD = ''.freeze

      STRING = [MAJOR, MINOR, PATCH, BUILD].compact.join('.').chomp('.')

      module_function

      def json_version
        {
          'version' => STRING
        }.to_json
      end
    end
  end
end


lib/kitchen/driver/marathon.rb
# -*- encoding: utf-8 -*-
#
# Author:: Anthony Spring (<aspring@yieldbot.com>)
#
# Copyright (C) 2016, Anthony Spring
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'deep_merge'
require 'json'
require 'kitchen'
require 'marathon'
require 'net/ssh'
require 'retryable'

module Kitchen
  module Driver
    # Marathon driver for Kitchen.
    #
    # @author Anthony Spring <aspring@yieldbot.com>
    class Marathon < Kitchen::Driver::SSHBase # rubocop:disable Metrics/ClassLength
      # Marathon Application Configuration

      default_config  :app_prefix,                'kitchen/'
      default_config  :app_template,              nil
      expand_path_for :app_template

      default_config  :app_config,                {}

      default_config  :app_launch_timeout,        30

      # Marathon HTTP Configuration

      default_config :marathon_host,              'http://localhost:8080'
      default_config :marathon_password,          nil
      default_config :marathon_username,          nil
      default_config :marathon_verify_ssl,        true

      # Marathon Proxy Configuration

      default_config :marathon_proxy_address,     nil
      default_config :marathon_proxy_port,        nil
      default_config :marathon_proxy_password,    nil
      default_config :marathon_proxy_username,    nil

      default_config(:instance_name) do |driver|
        driver.windows_os? ? nil : driver.instance.name
      end

      # Creates a new Driver object using the provided configuration data
      # which will be merged with any default configuration.
      #
      # @param config [Hash] provided driver configuration
      def initialize(config = {})
        # Let our parent do its work
        super(config)

        # Initialize marathon
        initialize_marathon
      end

      def create(state)
        return if state[:app_id]

        # Generate the application configuration
        app_config = generate_app_config

        # Create the app
        state[:app_id] = create_app(app_config)

        # Wait for the apps deployment to finish
        wait_for_app_deployment(app_config)

        # Update state
        update_app_state(state)
      end

      def converge(state)
        # Update the app state
        update_app_state(state)

        super(state)
      end

      def destroy(state)
        return if state[:app_id].nil?

        destroy_app(state[:app_id])

        state.delete(:app_id)
      end

      def setup(state)
        # Update the app state
        update_app_state(state)

        super(state)
      end

      def verify(state)
        # Update the app state
        update_app_state(state)

        super(state)
      end

      protected

      def create_app(app_config)
        Retryable.retryable(
          tries: 10,
          sleep: ->(n) { [2**n, 30].min },
          on: [::Marathon::Error::TimeoutError]
        ) do |_r, _|
          info("Creating the application: #{app_config['id']}")

          # Create the application
          ::Marathon::App.create(app_config)
        end

        app_config['id']
      end

      def create_app_id # rubocop:disable Metrics/LineLength
        # Need to remove any underscores from the app name
        "#{config[:app_prefix]}#{config[:instance_name]}-#{SecureRandom.hex}".tr('_', '-')
      end

      def destroy_app(app_id)
        ::Marathon::App.delete(app_id)
      rescue ::Marathon::Error::NotFoundError
        info("App (#{app_id}) not found.")
      end

      def generate_app_config # rubocop:disable Metrics/MethodLength
        # Generate the necessary config
        necessary_config = {}
        necessary_config['id']         = create_app_id
        necessary_config['instances']  = 1

        # Bring in the user defined JSON template
        user_config = if File.file?(config[:app_template])
                        JSON.parse(IO.read(config[:app_template]))
                      else
                        {}
                      end

        # Convert the app config into non-symbolized names
        app_config = JSON.parse(config[:app_config].to_json, symbolize_names: false)

        # Overlay the app config
        user_config.deep_merge!(app_config)

        # Overlay the necessary config
        user_config.deep_merge!(necessary_config)

        # Return the derived configuration
        user_config
      end

      def initialize_marathon
        # Initialize Marathon based off of configuration data
        marathon = {}

        # Basic HTTP Information
        marathon[:username]       = config[:marathon_username]
        marathon[:password]       = config[:marathon_password]

        # Basic SSL information
        marathon[:verify]         = config[:marathon_verify_ssl]

        # Basic Proxy information
        marathon[:http_proxyaddr] = config[:marathon_proxy_address]
        marathon[:http_proxyport] = config[:marathon_proxy_port]
        marathon[:http_proxyuser] = config[:marathon_proxy_username]
        marathon[:http_proxypass] = config[:marathon_proxy_password]

        # Set the Marathon credentials if given
        ::Marathon.options = marathon

        # Set the Marathon URL
        ::Marathon.url = config[:marathon_host]
      end

      def update_app_state(state) # rubocop:disable Metrics/MethodLength, Metrics/LineLength
        info('Refreshing host and port from Marathon...')

        app = nil

        # Get the host and port to SSH on
        Retryable.retryable(
          tries: 10,
          sleep: ->(n) { [2**n, 30].min },
          on: [::Marathon::Error::TimeoutError]
        ) do |_r, _|
          # Get the app
          app = ::Marathon::App.get(state[:app_id])
        end

        # Get the host
        state[:hostname] = app.info[:tasks][0][:host]

        # Get the mappings
        mappings = app.info[:container][:docker][:portMappings]

        # Get the index
        ssh_index = case mappings.length
        when 1
          info("Single port mapping found! Defaulting to that port.")

          # If there is a single mapping assume it is the SSH service port
          0
        else
          info("Multiple port mappings found! Looking for label 'SERVICE' => 'ssh'...")

          # Attempt to find the index via labels
          mappings.find_index(mappings.find { |mapping| mapping[:labels][:SERVICE] == 'ssh' })
        end

        info("Unable to determine SSH port index from mappings!") if ssh_index.nil?

        state[:port] = ssh_index.nil? ? 0 : app.info[:tasks][0][:ports][ssh_index]
      end

      def wait_for_app_deployment(app_config) # rubocop:disable Metrics/LineLength
        Retryable.retryable(
          tries: 10,
          sleep: ->(n) { [2**n, config[:app_launch_timeout]].min },
          on: [::Marathon::Error::TimeoutError, ::Timeout::Error]
        ) do |_r, _|
          info("Waiting for application to deploy: #{app_config['id']}")

          raise(::Timeout::Error.new, 'App is not running.') if ::Marathon::App.get(app_config['id']).info[:tasksRunning] == 0

          info("Application #{app_config['id']} is running.")
        end
      end
    end
  end
end
